{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c6faba29-f61c-41ec-a993-7431126cf1d3",
   "metadata": {},
   "source": [
    "## Co-culture interaction experiment | Part 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e52a9ecd-00c1-4c0d-a1c3-0242a0f6af67",
   "metadata": {},
   "source": [
    "<br>\n",
    "<b>Description</b> : In this notebook we run Intergram on the colon data.<br>\n",
    "<b>Author</b> : Alma Andersson (andera29@gene.com)<br>\n",
    "<b>Date</b> : 08/14/2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3574a3d-c5d2-48de-999c-50009555451f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32ea4895-bab0-4f2e-b6e7-b96b919d6731",
   "metadata": {},
   "source": [
    "Import relevant packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ba667b15-3230-4217-88a7-2143f848ae2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import telegraph as tg\n",
    "import anndata as ad \n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import os.path as osp\n",
    "import pandas as pd\n",
    "import xarray as xr\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "from scipy.stats import hypergeom"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1328ccb0-1d27-4ea1-9005-9b9f6d6448a7",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "20e98b0c-695a-4d63-b48b-bc642ab2b86a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_h5ad_uniqify(path,tag = None):\n",
    "    adata = ad.read_h5ad(path)\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    if tag is not None:\n",
    "        adata.obs_names  = [f'{tag}_{x}' for x in adata.obs_names]\n",
    "    return adata\n",
    "\n",
    "def pp_adata(ad_sc,ad_sp = None):\n",
    "    is_mt = ad_sc.var_names.str.startswith((\"mt-\",\"Mt\",\"ENSMUSG\"))\n",
    "    is_rp = ad_sc.var_names.str.startswith((\"rps\", \"rpl\",\"rp-\",\"rp\"))\n",
    "    is_rik = ad_sc.var_names.str.endswith((\"Rik\"))\n",
    "    keep_genes = (~is_mt) & (~is_rp) & (~is_rik)\n",
    "    ad_sc = ad_sc[:,keep_genes].copy()\n",
    "\n",
    "    sc.pp.filter_cells(ad_sc,min_counts=300)\n",
    "    \n",
    "    sc.pp.filter_genes(ad_sc,min_counts=10)\n",
    "\n",
    "    if ad_sp is not None:\n",
    "        sc.pp.filter_cells(ad_sp,min_counts=100)\n",
    "        sc.pp.filter_genes(ad_sp,min_counts=10)\n",
    "  \n",
    "\n",
    "    ad_sc.layers['raw'] = ad_sc.X.copy()\n",
    "    sc.pp.normalize_total(ad_sc,1e4)\n",
    "    sc.pp.log1p(ad_sc)\n",
    "    sc.pp.highly_variable_genes(ad_sc,n_top_genes=5000)\n",
    "    ad_sc.layers['norm'] = ad_sc.X.copy()\n",
    "    ad_sc.X = ad_sc.layers['raw'].copy()\n",
    "\n",
    "    if ad_sp is not None:\n",
    "        return ad_sc,ad_sp\n",
    "    return ad_sc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63db92ef-9d50-432b-8d9c-8928eaee655e",
   "metadata": {},
   "source": [
    "## Data and Processing\n",
    "Both the single cell and spatial data was downloaded from: https://cellxgene.cziscience.com/collections/4fa07e63-f712-4d8c-b885-2c515b5e2743.\n",
    "<br>\n",
    "For the scRNA-seq data we used the \"Global\" data object, for the spatial data we took all non anti_IL10r samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c628966c-9cbe-4d3e-ba28-fa00ac4bc225",
   "metadata": {},
   "source": [
    "Set data directory and filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fb5ac37b-e570-405a-9ddf-b450a009482b",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = '../data/imod/mouse_immune/'\n",
    "\n",
    "with open('OUTPUT_DIR.txt','r+') as f:\n",
    "    OUTPUT_ROOT = f.readlines()[0]\n",
    "\n",
    "OUTPUT_DIR = osp.join(OUTPUT_ROOT, 'coculture')\n",
    "os.makedirs(OUTPUT_DIR, exist_ok = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e483e3c6-001f-4323-8318-a9ba4b4f28de",
   "metadata": {},
   "outputs": [],
   "source": [
    "sp_files = ['1A_Hh.h5ad',\n",
    "'2A_Hh.h5ad',\n",
    "'3A_Hh.h5ad',\n",
    "'4A_Hh.h5ad',\n",
    "        ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87babea8-88fa-4a36-9fa4-2e56bc36d928",
   "metadata": {},
   "source": [
    "Load data and concatenate spatial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fd96bcf9-959f-4a99-a5e8-dc6a38d66c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_sp =  ad.concat([read_h5ad_uniqify(osp.join(DATA_DIR,fn)) for fn in sp_files])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cbbe8009-6874-4cc3-84c2-99074ed52228",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_sp = ad_sp[ad_sp.obs['in_tissue'].values == 1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f05cd6e-133a-4659-a0f2-336bdf8d644d",
   "metadata": {},
   "source": [
    "Load single cell data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fbe4b47e-4925-4aed-bd5e-4df54950cefd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_sc = ad.read_h5ad(osp.join(DATA_DIR,'sc.h5ad'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33df7ad3-6dfc-47a8-8deb-0f75b8291577",
   "metadata": {},
   "source": [
    "Convert ENSEMBL IDs to Gene Names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "167d5376-fccd-4344-9001-62799f735371",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensg_to_hgnc = ad_sc.var[['feature_name']].copy()\n",
    "ad_sc.var_names = ensg_to_hgnc.loc[ad_sc.var_names].values.flatten()\n",
    "inter = ad_sp.var_names.intersection(ensg_to_hgnc.index)\n",
    "ad_sp = ad_sp[:,inter]\n",
    "ad_sp.var_names = ensg_to_hgnc.loc[ad_sp.var_names].values.flatten()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a700fed2-8e2f-4708-ac27-4b7bb0a77c31",
   "metadata": {},
   "source": [
    "We want to look at the interaction between CD4+ T-cells and Dendritic cells, so we'll merge some of the finer subtypes into a single T respecitevey DC class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2fbf6847-f4c8-4d39-8553-fc7ca8270165",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_t_cells = [\n",
    "    'naive thymus-derived CD4-positive, alpha-beta T cell',\n",
    "    'regulatory T cell',\n",
    "    'CD4-positive, alpha-beta memory T cell',\n",
    "    'T-helper 17 cell'\n",
    "]\n",
    "\n",
    "cd11c_dendritic_cells = [\n",
    "    'conventional dendritic cell',\n",
    "    'CD103-positive dendritic cell',\n",
    "    'CD8_alpha-positive CD11b-negative dendritic cell',\n",
    "    'dendritic cell'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2e6c7c00-7c88-4374-a3d3-589f33703e37",
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_map = {key:key for key in np.unique(ad_sc.obs['cell_type'].values) }\n",
    "ct_map = {key:(val  if key not in cd4_t_cells else 'Tcell_CD4+') for key,val in ct_map.items() }\n",
    "ct_map = {key:(val  if key not in cd11c_dendritic_cells else 'DC') for key,val in ct_map.items() }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c81375fa-c788-4c63-a710-9f6113c4e7ed",
   "metadata": {},
   "source": [
    "Add the curated cell type labels to the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6489b4e1-9855-4321-af44-ea7417017338",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_sc.obs['curated_celltypes'] = ad_sc.obs['cell_type'].map(ct_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93df16e4-23ab-4817-87b7-ae7568a45226",
   "metadata": {},
   "source": [
    "## Run Mapping Method and Interaction Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8499e9f8-ee94-4718-93ab-d89d5be0c285",
   "metadata": {},
   "source": [
    "Define label to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f30e92b1-24b8-4f89-846d-cd1daddb0118",
   "metadata": {},
   "outputs": [],
   "source": [
    "label_col = 'curated_celltypes'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b372d0e7-b9b0-40d6-a6f5-30cb5102b2fe",
   "metadata": {},
   "source": [
    "Preprocess data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c3f07e28-8d78-43e1-a362-1c64b4ddfc46",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_sc,ad_sp = pp_adata(ad_sc,ad_sp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58b5d2ea-2f77-4c48-a513-75bc439a7c07",
   "metadata": {},
   "source": [
    "Prepare `state_dict` input for telegraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75df627f-2e01-4bb4-964a-564aacc17d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg_genes = ad_sc.var_names[ad_sc.var.highly_variable.values].tolist()\n",
    "\n",
    "input_dict_1 = tg.met.utils.adatas_to_input({'from':ad_sc, 'to':ad_sp}, # provide the data to be used\n",
    "                                                    categorical_labels={'from':[label_col]}, # include cluster labels in the design matrix\n",
    "                                                  )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86602da5-7b4c-4559-abcd-7cf48fed1ead",
   "metadata": {},
   "source": [
    "Run mapping method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "85298a27-6a0f-4e1f-88d5-54d3a620554f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 4489 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set Solid Seed\n",
      "Set Solid Seed\n",
      "Score: 0.557, KL reg: 3.431, Entropy reg: -12.214\n",
      "Score: 0.705, KL reg: 3.219, Entropy reg: -9.825\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.816\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.809\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.805\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.803\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.802\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.802\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.802\n",
      "Score: 0.706, KL reg: 3.219, Entropy reg: -9.801\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Renormalizing Single cell data\n",
      "INFO:root:Begin training with 4489 genes and rna_count_based density_prior in cells mode after renormalization\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set Solid Seed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set Solid Seed\n",
      "Score: 0.690, KL reg: 0.182, Entropy reg: -97380.273\n",
      "Score: 0.769, KL reg: 0.002, Entropy reg: -88910.836\n",
      "Score: 0.775, KL reg: 0.001, Entropy reg: -77517.508\n",
      "Score: 0.777, KL reg: 0.000, Entropy reg: -68511.109\n",
      "Score: 0.778, KL reg: 0.000, Entropy reg: -64167.742\n",
      "Score: 0.779, KL reg: 0.000, Entropy reg: -60468.164\n",
      "Score: 0.779, KL reg: 0.000, Entropy reg: -56704.879\n",
      "Score: 0.780, KL reg: 0.000, Entropy reg: -53021.523\n",
      "Score: 0.780, KL reg: 0.000, Entropy reg: -49411.641\n",
      "Score: 0.780, KL reg: 0.000, Entropy reg: -46150.680\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n"
     ]
    }
   ],
   "source": [
    "\n",
    "tg.met.pp.StandardTangramV2.run(input_dict_1)\n",
    "\n",
    "map_res_1 = tg.met.map_methods.TangramV2Map.run(input_dict_1,\n",
    "                                                num_epochs = 1000,\n",
    "                                                genes =hvg_genes,\n",
    "                                             )\n",
    "\n",
    "input_dict_1.update(map_res_1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ca4b08-ba79-4e10-ae15-b1813a2ecb90",
   "metadata": {},
   "source": [
    "Run Interaction Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f9d666eb-25b4-49eb-86b4-9d85dd4dc1f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "HPU available: False, using: 0 HPUs\n",
      "You are using a CUDA device ('NVIDIA A100-SXM4-80GB') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n",
      "\n",
      "  | Name         | Type | Params | Mode\n",
      "---------------------------------------------\n",
      "  | other params | n/a  | 326 K  | n/a \n",
      "---------------------------------------------\n",
      "326 K     Trainable params\n",
      "0         Non-trainable params\n",
      "326 K     Total params\n",
      "1.304     Total estimated model params size (MB)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "abae7f7414c649aca48599b0050d4489",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Training: |                                                                                                   â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`Trainer.fit` stopped: `max_epochs=1000` reached.\n"
     ]
    }
   ],
   "source": [
    "tg.met.pp.StandardScanpy.run(input_dict_1,target_objs = ['X_from'])\n",
    "\n",
    "inter_res = tg.dev.imod.methods.InteractionModel.run(input_dict_1,\n",
    "                                                     n_epochs = 1000,\n",
    "                                                     learning_rate = 0.01)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "05dac187-0ace-4be3-9537-f83d0bbc1b98",
   "metadata": {},
   "outputs": [],
   "source": [
    "inter_res.to_netcdf(osp.join(OUTPUT_DIR,'inter_res.netcdf'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c70ec64c-d0ee-4130-a85a-e6e8ddd98bf0",
   "metadata": {},
   "source": [
    "## Downstream Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dea70d3-d202-4c85-93e8-97099e7d2b5d",
   "metadata": {},
   "source": [
    "Get model coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8431c8f4-bc90-4ae1-a061-42ce755eea5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "label = 'DC_scanpy'\n",
    "corr_res = dict()\n",
    "overlap_res = dict()\n",
    "alpha_sig = 0.05\n",
    "\n",
    "# iterate over each interaction\n",
    "for inter in beta.index.values:\n",
    "\n",
    "    # get beta coefs. for given interaction\n",
    "    pred_df = pd.DataFrame(beta.loc[inter]['beta'])\n",
    "    pred_df.columns = ['beta']\n",
    "\n",
    "    # get the reference genes\n",
    "    true_df = true[label].copy()\n",
    "\n",
    "    # only keep genes that were present in the sc data\n",
    "    loc_inter = true_df.index.intersection(pred_df.index)\n",
    "\n",
    "    # make sure pred and reference share genes\n",
    "    pred_df = pred_df.loc[loc_inter]\n",
    "    true_df = true_df.loc[loc_inter]\n",
    "\n",
    "    # only keep the reference genes with a significant p-value\n",
    "    true_df = true_df.iloc[true_df.pvals_adj.values <= alpha_sig]\n",
    "    # get up-regulated reference genes\n",
    "    true_df = true_df[true_df.logfoldchanges.values > 1]\n",
    "\n",
    "    # sort the predicted and reference values\n",
    "    top_pred =  pred_df.sort_values('beta',ascending = False).index\n",
    "    top_true =  true_df.sort_values('logfoldchanges',ascending = False).index\n",
    "\n",
    "    # compare the overlap\n",
    "    overlap =  top_true.intersection(top_pred[0:len(top_true)])\n",
    "\n",
    "    # register results\n",
    "    overlap_res[inter] = dict(n_overlap = len(overlap),\n",
    "                              n_pred =len(top_true),\n",
    "                              n_true = len(top_true),\n",
    "                              n_background = len(loc_inter),\n",
    "                              overlap = overlap,\n",
    "                              top_pred = top_pred,\n",
    "                              top_true = top_true,\n",
    "                              background = loc_inter )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5211c2a6-a6e3-4e57-a50c-189671790258",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj = overlap_res[sel_inter]\n",
    "pval = hypergeometric_test(**obj)\n",
    "top_number = len(obj['top_true'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1c6600aa-d966-4d5e-b2b9-8556284c431b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[IM] OVERLAP: 387/863 | PVAL: 9.62E-274\n"
     ]
    }
   ],
   "source": [
    "print('[IM] OVERLAP: {}/{} | PVAL: {:0.2E}'.format(obj['n_overlap'],top_number,pval))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4325dc69-0aff-4df3-99ae-3cc7f4d0e6f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get beta coefs. for given interaction\n",
    "\n",
    "for inter_name in ['DC_vs_Tcell_CD4+','Tcell_CD4+_vs_DC']:\n",
    "    pred_df = pd.DataFrame(beta.loc[inter_name]['beta'])\n",
    "    pred_df.columns = ['beta']\n",
    "    pred_df.to_csv(osp.join(OUTPUT_DIR, 'beta_{}.csv'.format(inter_name)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b47b0bc-794b-41eb-918b-f4f9ba10092d",
   "metadata": {},
   "source": [
    "Compare with Highly Variable Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a3f1bd03-43ef-4af4-9316-8530246e12e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[HVG] OVERLAP: 180/863 | PVAL: 1.36E-49\n"
     ]
    }
   ],
   "source": [
    "pred_df = tg.methods.dea.HVGFeatureDEA.run_with_adata(input_dict_1['X_from'],\n",
    "                                                      subset_col=label_col, \n",
    "                                                      subset_labels=('DC' if label.startswith('DC') else 'Tcell_CD4+'))\n",
    "\n",
    "pred_df.index =  pd.Index([x.lower() for x in pred_df.index])\n",
    "\n",
    "pred_df.to_csv(osp.join(OUTPUT_DIR,'hvg_{}.csv'.format(label.split('_')[0])))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "telegraph",
   "language": "python",
   "name": "telegraph"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
