{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "001e5b04-67c3-4f50-97e7-6fce5908d86f",
   "metadata": {},
   "source": [
    "## B cell activation experiment\n",
    "<br>\n",
    "<b>Description</b> : In this notebook we run Tangram2 and Intergram on human lymph node data.<br>\n",
    "<b>Author</b> : Alma Andersson (andera29@gene.com)<br>\n",
    "<b>Date</b> : 08/14/2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0900002a-847c-4402-8f6b-0aaeea46294f",
   "metadata": {},
   "source": [
    "Import necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "99847a65-c7c4-4839-b97f-f249c2c584d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext blackcellmagic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5960941-330e-4ca9-b67d-5b1b81455d76",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/scratchfs01/site/u/andera29/micromamba/envs/tangram2/lib/python3.11/site-packages/louvain/__init__.py:54: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  from pkg_resources import get_distribution, DistributionNotFound\n"
     ]
    }
   ],
   "source": [
    "import tangram2 as tg2\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import hypergeom\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy\n",
    "import tqdm\n",
    "import os.path as osp\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93f6b6dd-8e6c-4442-9e4d-da2b5ffc21ac",
   "metadata": {},
   "source": [
    "## Define Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7be301d9-74ae-41e5-b227-b4abb20eb230",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def get_isoforms(s):\n",
    "    pattern = r'.*\\.\\d+'\n",
    "    return bool(re.fullmatch(pattern, s))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5dddeb2-22a8-4928-8ce6-99dcefa84a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pp_adata(ad_sc,ad_sp = None):\n",
    "    drop1 = ad_sc.var_names.str.startswith((\"MT-\",\"MT\",\"RP\",\"RPS\",\"LINC\"))\n",
    "    drop2 = ad_sc.var_names.str.endswith(('VDJsum'))\n",
    "    drop3 = np.array([ get_isoforms(x) for x in ad_sc.var_names])\n",
    "    ad_sc = ad_sc[:,(~drop1) & (~drop2) & (~drop3)].copy()\n",
    "\n",
    "    sc.pp.filter_cells(ad_sc,min_counts=300)\n",
    "    \n",
    "    sc.pp.filter_genes(ad_sc,min_counts=10)\n",
    "\n",
    "    if ad_sp is not None:\n",
    "        sc.pp.filter_cells(ad_sp,min_counts=100)\n",
    "        sc.pp.filter_genes(ad_sp,min_counts=10)\n",
    "  \n",
    "\n",
    "    ad_sc.layers['raw'] = ad_sc.X.copy()\n",
    "    sc.pp.normalize_total(ad_sc,1e4)\n",
    "    sc.pp.log1p(ad_sc)\n",
    "    sc.pp.highly_variable_genes(ad_sc,n_top_genes=5000)\n",
    "    ad_sc.layers['norm'] = ad_sc.X.copy()\n",
    "    ad_sc.X = ad_sc.layers['raw'].copy()\n",
    "\n",
    "    if ad_sp is not None:\n",
    "        return ad_sc,ad_sp\n",
    "    return ad_sc\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a5df9d80-7e53-475c-958b-3d49776dfcf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('OUTPUT_DIR.txt','r+') as f:\n",
    "    OUTPUT_ROOT = f.readlines()[0]\n",
    "\n",
    "OUTPUT_DIR = osp.join(OUTPUT_ROOT, 'bcell')\n",
    "os.makedirs(OUTPUT_DIR, exist_ok = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99b69721-840b-4dd8-9dc1-4fe0c62a2a69",
   "metadata": {},
   "source": [
    "### Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb5d8b28-bd40-445f-92dd-887e455ccb1c",
   "metadata": {},
   "source": [
    "Get data and process <br>\n",
    "From: https://github.com/BayraktarLab/cell2location/blob/master/docs/notebooks/cell2location_tutorial.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3d25cfc-f59e-4af5-8e89-c8c7c8c0ed3d",
   "metadata": {},
   "source": [
    "Spatial Visium data from the human lymph node, all processing follows the author provided tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a436fa49-e43b-4e31-bf71-05925f382840",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_vis = sc.datasets.visium_sge(sample_id=\"V1_Human_Lymph_Node\")\n",
    "adata_vis.obs['sample'] = list(adata_vis.uns['spatial'].keys())[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b92ccfa3-1659-497b-b124-b5d70c87cbc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_vis.var['SYMBOL'] = adata_vis.var_names\n",
    "adata_vis.var.set_index('gene_ids', drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6c022407-9ebe-4a65-9329-c7f963854730",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_vis.var.index = adata_vis.var['SYMBOL']\n",
    "adata_vis.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cb4f8b8a-44b5-4a1c-8322-f1b5a8aeb426",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find mitochondria-encoded (MT) genes\n",
    "adata_vis.var['MT_gene'] = [gene.startswith('MT-') for gene in adata_vis.var['SYMBOL']]\n",
    "    \n",
    "# remove MT genes for spatial mapping (keeping their counts in the object)\n",
    "adata_vis.obsm['MT'] = adata_vis[:, adata_vis.var['MT_gene'].values].X.toarray()\n",
    "adata_vis = adata_vis[:, ~adata_vis.var['MT_gene'].values]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c805ff4-8cfd-4ebf-acfc-149acdc3d9c1",
   "metadata": {},
   "source": [
    "Get paired single cell data (originally downloaded from: https://cell2location.cog.sanger.ac.uk/paper/integrated_lymphoid_organ_scrna/RegressionNBV4Torch_57covariates_73260cells_10237genes/sc.h5ad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2e8b68e5-5b55-42b8-9bb2-56849ed69362",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Read data\n",
    "adata_ref = sc.read(\n",
    "    f'../../data/lymph_node/lymph_sc.h5ad', \n",
    "    backup_url='https://cell2location.cog.sanger.ac.uk/paper/integrated_lymphoid_organ_scrna/RegressionNBV4Torch_57covariates_73260cells_10237genes/sc.h5ad'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "61e5c187-911c-43c1-9d52-a60e3c69267b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ref.var['SYMBOL'] = adata_ref.var.index \n",
    "# rename 'GeneID-2' as necessary for your data\n",
    "adata_ref.var.set_index('GeneID-2', drop=True, inplace=True)\n",
    "\n",
    "# delete unnecessary raw slot (to be removed in a future version of the tutorial)\n",
    "del adata_ref.raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7db1d812-ca8c-4698-ae7d-dab90543a91b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ref.var.index = adata_ref.var['SYMBOL']\n",
    "adata_ref.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0b23d6d-6716-4a8e-8ab9-4fc96fd607d7",
   "metadata": {},
   "source": [
    "We'll only use one donor for the scRNA-seq data to demonstrate (it's also less computationally heavy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e286dbac-8d96-43ce-9cb6-191ea27c312b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ref = adata_ref[adata_ref.obs['Donor'].values == 'A16'].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12037ab4-808a-4ef1-9bca-ccd56e767e40",
   "metadata": {},
   "source": [
    "## Process data and run workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "be917bc6-45a6-4646-b268-19fe2cd13345",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_ref,adata_vis = pp_adata(adata_ref,adata_vis)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40ae6286-5216-4f86-9e9a-41d69c92a64e",
   "metadata": {},
   "source": [
    "Define column with cell type labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ec8436ff-3dfb-4e18-b1b8-efb89432b0d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "label_col = 'Subset_Broad'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72e91476-7174-4cda-8345-8cfa58fa9a19",
   "metadata": {},
   "source": [
    "Initialize state dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bed24554-78ce-466b-a287-69d59dbc63fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg_genes = adata_ref.var_names[adata_ref.var.highly_variable.values].tolist()\n",
    "\n",
    "input_dict_1 = tg2.evalkit.met.utils.adatas_to_input({'from':adata_ref,\n",
    "                                             'to':adata_vis}, # provide the data to be used\n",
    "                                             categorical_labels={'from':[label_col]}, # include cluster labels in the design matrix\n",
    "                                            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e4bd13-3bb1-4f07-9e4d-ea65fe166309",
   "metadata": {},
   "source": [
    "Run mapping method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e35933a8-ea72-47ee-995f-b4bdf7ec3773",
   "metadata": {},
   "outputs": [],
   "source": [
    "tg2.evalkit.met.pp.StandardTangram2.run(input_dict_1)\n",
    "\n",
    "map_res_1 = tg2.evalkit.met.map_methods.Tangram2Map.run(\n",
    "    input_dict_1,\n",
    "    num_epochs=1000,\n",
    "    genes=hvg_genes,\n",
    ")\n",
    "\n",
    "input_dict_1.update(map_res_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "086eb2c2-ab48-4e4f-949f-003033162838",
   "metadata": {},
   "source": [
    "Run interaction model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cca0dd05-9481-4e72-b81c-4e753d1d15ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "tg2.evalkit.met.pp.StandardScanpy.run(input_dict_1, target_objs=[\"X_from\"])\n",
    "\n",
    "inter_res = tg2.ccc.TangramCCC.run(input_dict_1, n_epochs=1000, learning_rate=0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "51d503f2-2924-408f-aeda-5bea03f2c760",
   "metadata": {},
   "outputs": [],
   "source": [
    "inter_res.to_netcdf(osp.join(OUTPUT_DIR,'inter_res.netcdf'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48e4300a-51d2-4e04-873e-e73da6ea04ae",
   "metadata": {},
   "source": [
    "Calculate highly variable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6a3b1310-20be-475d-a192-6538243f3d08",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg = tg2.evalkit.methods.dea.HVGFeatureDEA.run_with_adata(\n",
    "    input_dict_1[\"X_from\"],\n",
    "    subset_col=label_col,\n",
    "    subset_labels=\"B_GC_prePB\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4922331b-adb1-461a-be47-434a8f4bfebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg.to_csv(osp.join(OUTPUT_DIR,'hvg_B_GC_prePB.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tangram2",
   "language": "python",
   "name": "tangram2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
