{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cc14584c-5b96-49b9-8a41-bf5f06ad49ba",
   "metadata": {},
   "source": [
    "## Immunosuppressive paths between cancer epithelial and CD8 T cells in BC datasets\n",
    "<br>\n",
    "<b>Description</b> : In this notebook we tried to generate the perfectly paired, paired and unpaired datasets between single cell and spatial transcriptomics for benchmarking purpose in Figure 2c.<br>\n",
    "<b>Author</b> : Hejin Huang (huang.hejin@gene.com)<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6ccbb2fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/raid/home/huangh83/miniforge3/envs/tangram2/lib/python3.11/site-packages/louvain/__init__.py:54: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  from pkg_resources import get_distribution, DistributionNotFound\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import tangram2 as tg2\n",
    "import tangram as tg\n",
    "from sklearn.metrics import jaccard_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ce6b1744-fe2c-45a8-916a-2556b8365a7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Data Loading and Initial Filtering ---\n",
    "path = '../../data/tangram2_paper_data/original/bc/sc/'\n",
    "ad_sc = sc.read_h5ad(path + 'bc.h5ad')\n",
    "\n",
    "subtype = 'TNBC'\n",
    "ad_sc_TNBC = ad_sc[ad_sc.obs['subtype'] == subtype].copy() # Use .copy() to avoid SettingWithCopyWarning\n",
    "\n",
    "patient_IDs = list(ad_sc_TNBC.obs['orig.ident'].unique())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2613b2dc-7248-4023-85be-beb8c903033f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing patient: CID4495 with 7985 cells.\n",
      "Processing patient: CID44971 with 7986 cells.\n",
      "Processing patient: CID44991 with 7023 cells.\n",
      "Processing patient: CID4513 with 5619 cells.\n",
      "Processing patient: CID4515 with 4149 cells.\n",
      "Processing patient: CID3963 with 3527 cells.\n"
     ]
    }
   ],
   "source": [
    "# --- Data Generation Loop for Patients with > 3000 cells ---\n",
    "output_base_path = '../../data/tangram2_paper_data/imod/bc/simulated_dataset/'\n",
    "label_used = 'celltype_major' # Define label column once\n",
    "\n",
    "for patient in patient_IDs:\n",
    "    ad_sc_patient = ad_sc_TNBC[ad_sc_TNBC.obs['orig.ident'] == patient].copy()\n",
    "\n",
    "    # Check if the patient has more than 3000 cells\n",
    "    if len(ad_sc_patient) > 3000:\n",
    "        print(f\"Processing patient: {patient} with {len(ad_sc_patient)} cells.\")\n",
    "\n",
    "        # Create patient-specific directory if it doesn't exist\n",
    "        patient_output_dir = os.path.join(output_base_path, patient)\n",
    "        os.makedirs(patient_output_dir, exist_ok=True)\n",
    "\n",
    "        # Generate 'random' column with 0 or 1 (50% probability each)\n",
    "        # Using 1 and 2 to match the original notebook's splitting logic\n",
    "        ad_sc_patient.obs['random'] = np.random.randint(1, 3, size=len(ad_sc_patient))\n",
    "\n",
    "        ad_sc_patient_A1 = ad_sc_patient[ad_sc_patient.obs['random'] == 1].copy()\n",
    "        ad_sc_paired_unperfect = ad_sc_patient[ad_sc_patient.obs['random'] == 2].copy()\n",
    "\n",
    "        # Generate paired dataset using cellmix\n",
    "        ad_sp, ad_sc_paired = tg2.evalkit.datagen.cellmix.cellmix.cellmix(\n",
    "            ad_sc_patient_A1,\n",
    "            n_spots=100,\n",
    "            n_cells_per_spot=10,\n",
    "            n_types_per_spot=3,\n",
    "            label_col=label_used,\n",
    "            encode_spatial=True,\n",
    "        )\n",
    "\n",
    "        # Save the generated paired and unperfectly paired datasets\n",
    "        # ad_sc_paired.write_h5ad(os.path.join(patient_output_dir, 'ad_sc_paired.h5ad'))\n",
    "        # ad_sc_paired_unperfect.write_h5ad(os.path.join(patient_output_dir, 'ad_sc_paired_unperfect.h5ad'))\n",
    "        # ad_sp.write_h5ad(os.path.join(patient_output_dir, 'ad_sp.h5ad')) # Save ad_sp as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b01a737-b590-4c52-85e8-83ab6ec567de",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76d3fdcc-55a3-49a5-b293-d90cd60c3440",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c89a1d7e-6311-419d-a54f-5e9f5bcbdf19",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
