{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e6e6d0b-f33b-4c32-bc31-76de0055a707",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f18d49c0-635e-4cbb-a5fe-380607d910b7",
   "metadata": {},
   "source": [
    "Import necessary standard packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9063cb70-552b-4639-83ad-61dd4566ca18",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "import subprocess\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5bca103-b3e7-44ec-87d2-4a209140e380",
   "metadata": {},
   "source": [
    "Import evaluation module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8da4e0c2-cb98-4804-86e9-f400fd63dc06",
   "metadata": {},
   "outputs": [],
   "source": [
    "EVAL_PTH = '../src/'\n",
    "import sys\n",
    "if EVAL_PTH not in sys.path:\n",
    "    sys.path.append(EVAL_PTH)\n",
    "import eval as ev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f7d4fcb2-baf5-425d-af40-a232da24760a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "methods:\n",
      "  - map_random\n",
      "  - map_max_corr\n",
      "  - map_tangram_v1\n",
      "  - map_tangram_v2\n",
      "  - map_CeLEry\n",
      "  - prd_tangram_v1\n",
      "  - prd_tangram_v2\n",
      "  - grp_threshold\n",
      "  - dea_scanpy\n",
      "metrics:\n",
      "  - map_jaccard\n",
      "  - map_accuracy\n",
      "  - map_rmse\n",
      "  - dev_print\n",
      "preprocess:\n",
      "  - standard_scanpy\n",
      "  - normalize_totaly\n",
      "  - CeLEry\n",
      "  - tangramv1\n",
      "  - tangramv2\n"
     ]
    }
   ],
   "source": [
    "# Listing all the available methods, etc.\n",
    "ev.info.list_methods()\n",
    "ev.info.list_metrics()\n",
    "ev.info.list_pp()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b33b445-2ff8-4ee1-b547-2871e3a529b8",
   "metadata": {},
   "source": [
    "Specify path to config file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2c8aecbd-2bab-442c-be8f-f38df9f9127e",
   "metadata": {},
   "outputs": [],
   "source": [
    "CONFIG_PTH = 'hejinwf_config.yaml'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7999c3d-02ae-4047-866e-2858b6627aa5",
   "metadata": {},
   "source": [
    "Inspect config file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7ec0d8a3-baa7-44b2-8423-cc9be3eb0392",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---\n",
      "\n",
      "data:\n",
      "  exp1:\n",
      "    sp:\n",
      "      path: '/gstore/data/resbioai/tangram2_data/breast_cancer/spatial/ad_sp_CID44971.h5ad'\n",
      "    sc:\n",
      "      path: '/gstore/data/resbioai/tangram2_data/breast_cancer/single_cell/Wu_etal_2021_BRCA_scRNASeq/ad_sc.h5ad'\n",
      "\n",
      "methods:\n",
      "  exp1:\n",
      "    - wf_hejin\n",
      "\n",
      "method_params:\n",
      "  exp1:\n",
      "    wf_hejin:\n",
      "      map_args:\n",
      "        num_epochs: 1000\n",
      "      group_args:\n",
      "        feature_name: 'cd274'\n",
      "        thres_x: 9\n",
      "\n",
      "\n",
      "preprocess:\n",
      "  exp1:\n",
      "    sc:\n",
      "      total_normalize:\n",
      "        target_sum: 1e4\n",
      "metrics:\n",
      "  exp1:\n",
      "    wf_hejin:\n",
      "      dev_print: \n"
     ]
    }
   ],
   "source": [
    "!cat $CONFIG_PTH"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d94e0457-d36a-49aa-a9ff-80956c15653e",
   "metadata": {},
   "source": [
    "Read config file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b63f935-3c7f-4273-b6e5-614c7022e697",
   "metadata": {},
   "source": [
    "we'll use a \"stupid\" metric that just prints the output, since this is for illustrative purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2b0706d0-3c6c-4cfd-9d15-7e8ba612b7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "\n",
    "with open(CONFIG_PTH, 'r') as f:\n",
    "    design = yaml.safe_load(f)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50cfc03b-1203-43dd-a4b7-73d25cae1f81",
   "metadata": {},
   "source": [
    "Run evaluation and look at output (saved to temporary folder for demo purpose, will be removed once context closes).\n",
    "<br>\n",
    "<br>\n",
    "**Note**: if you actually want to inspect the results, change the context from a `TemporaryDirectory` to any path of your choice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6da8b6ad-2e7b-49ce-9009-6221d5393cb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:16095 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:16095 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 16095 genes and rna_count_based density_prior in cells mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.768, VG reg: 0.768, KL reg: 0.255\n"
     ]
    }
   ],
   "source": [
    "with tempfile.TemporaryDirectory() as tmpdir:\n",
    "    ev.run.run(design,tmpdir)\n",
    "\n",
    "    subprocess.run(['tree',tmpdir])\n",
    "    res_files = list(Path(tmpdir).rglob('*.txt'))\n",
    "    for fn in res_files:\n",
    "        print('\\n Content of: {}'.format(fn))\n",
    "        subprocess.run(['cat',fn])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f2ec846-0bb5-42d1-a987-563b602da912",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:tangram]",
   "language": "python",
   "name": "conda-env-tangram-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
